File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/json/schema/SchemaValidatorTest.java
package com.fincity.nocode.kirun.engine.json.schema;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.junit.jupiter.api.Test;

import com.fincity.nocode.kirun.engine.HybridRepository;
import com.fincity.nocode.kirun.engine.Repository;
import com.fincity.nocode.kirun.engine.json.schema.object.AdditionalPropertiesType;
import com.fincity.nocode.kirun.engine.json.schema.type.SchemaType;
import com.fincity.nocode.kirun.engine.json.schema.type.Type;
import com.fincity.nocode.kirun.engine.json.schema.validator.SchemaValidator;
import com.fincity.nocode.kirun.engine.json.schema.validator.exception.SchemaValidationException;
import com.fincity.nocode.kirun.engine.repository.KIRunSchemaRepository;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

class SchemaValidatorTest {

	@Test
	void schemaValidatortestForNullElement() {
		JsonObject defaultValue = new JsonObject();
		defaultValue.addProperty("value", 123);

		Schema schema = new Schema();
		schema.setType(Type.of(SchemaType.ARRAY));
		schema.setDefaultValue(defaultValue);

		JsonElement element = null;

		assertEquals(SchemaValidator.validate(null, schema, null, element), schema.getDefaultValue());

	}

	@Test
	void schemaValidatortestIfConstantNotEqualsElement() {
		JsonObject defaultValue = new JsonObject();
		defaultValue.addProperty("value", 123);

		JsonObject constantElement = new JsonObject();
		constantElement.addProperty("value", "constant");

		Schema schema = new Schema();
		schema.setType(Type.of(SchemaType.ARRAY));
		schema.setDefaultValue(defaultValue);
		schema.setConstant(constantElement);

		JsonElement element = new JsonObject();

		SchemaValidationException schemaValidationException = assertThrows(SchemaValidationException.class,
		        () -> SchemaValidator.validate(null, schema, null, element));

		assertEquals("null - Expecting a constant value : " + element, schemaValidationException.getMessage());

	}

	@Test
	void schemaValidatortestIfConstantEqualsElement() {
		JsonObject defaultValue = new JsonObject();
		defaultValue.addProperty("value", 123);

		JsonObject element = new JsonObject();
		element.addProperty("value", "constant");

		Schema schema = new Schema();
		schema.setType(Type.of(SchemaType.ARRAY));
		schema.setDefaultValue(defaultValue);
		schema.setConstant(element);

		assertEquals(SchemaValidator.validate(null, schema, null, element), element);

	}

	@Test
	void schemaValidatorTestForRefOfRef() {
		var locationMap = new HashMap<String, Schema>();
		var schemaMap = new HashMap<String, Schema>();
		locationMap.put("url", Schema.ofString("url"));
		var locationSchema = Schema.ofObject("Location")
		        .setNamespace("Test")
		        .setProperties(locationMap);
		var urlParamsSchema = Schema.ofObject("UrlParameters")
		        .setNamespace("Test")
		        .setAdditionalProperties(new AdditionalPropertiesType().setSchemaValue(Schema.ofRef("Test.Location")));
		var testSchema = Schema.ofObject("TestSchema")
		        .setNamespace("Test")
		        .setAdditionalProperties(
		                new AdditionalPropertiesType().setSchemaValue(Schema.ofRef("Test.UrlParameters")));
		schemaMap.put("Location", locationSchema);
		schemaMap.put("UrlParameters", urlParamsSchema);
		schemaMap.put("TestSchema", testSchema);
		class TestRepository implements Repository<Schema> {

			@Override
			public Schema find(String namespace, String name) {
				if (namespace == null) {
					return null;
				}
				return schemaMap.get(name);
			}

			@Override
			public List<String> filter(String name) {

				return schemaMap.values()
				        .stream()
				        .map(Schema::getFullName)
				        .filter(e -> e.toLowerCase()
				                .contains(name.toLowerCase()))
				        .toList();
			}
		}
		var repo = new HybridRepository<Schema>(new TestRepository(), new KIRunSchemaRepository());
		var urlParams = new JsonObject();
		var testValue = new JsonObject();
		var location = new JsonObject();
		location.addProperty("url", "http://test/");
		urlParams.add("obj", location);
		testValue.add("obj", urlParams);

		assertEquals(SchemaValidator.validate(null, Schema.ofRef("Test.TestSchema"), repo, testValue), testValue);

	}

	@Test
	void schemaValidatorTestForEnumCheck() {

		JsonObject defaultValue = new JsonObject();
		defaultValue.addProperty("value", 123);

		JsonArray element = new JsonArray();
		element.add(1);
		element.add(2);

		List<JsonElement> enums = new ArrayList<JsonElement>();
		enums.add(defaultValue);

		Schema schema = new Schema();
		schema.setType(Type.of(SchemaType.ARRAY));
		schema.setDefaultValue(defaultValue);
		schema.setEnums(enums);

		schema.setRef("test_ref");
		schema.setNot(schema);

		SchemaValidationException schemaValidationException = assertThrows(SchemaValidationException.class,
		        () -> SchemaValidator.validate(null, schema, null, element));

		assertEquals("null - Value is not one of " + schema.getEnums(), schemaValidationException.getMessage());

		// For positive case
		enums.add(element);
		schema.setEnums(enums);
		assertEquals(SchemaValidator.validate(null, schema, null, element), element);
	}

	@Test
	void schemaValidatorIfGetNot() {

		JsonObject defaultValue = new JsonObject();
		defaultValue.addProperty("value", 123);

		JsonArray element = new JsonArray();
		element.add(1);
		element.add(2);

		Schema schema = new Schema();
		schema.setDefaultValue(defaultValue);

		Schema setNotSchema = new Schema();
		schema.setNot(setNotSchema);

		SchemaValidationException schemaValidationException = assertThrows(SchemaValidationException.class,
		        () -> SchemaValidator.validate(null, schema, null, element));
		assertEquals("null.null - Schema validated value in not condition.", schemaValidationException.getMessage());

	}

	@Test
	void schemaValidatorForTypeValidation() {

		JsonObject defaultValue = new JsonObject();
		defaultValue.addProperty("value", 123);

		JsonArray element = new JsonArray();
		element.add(1);
		element.add(2);

		Schema schema = new Schema();
		schema.setDefaultValue(defaultValue);
		schema.setType(Type.of(SchemaType.ARRAY));

		assertEquals(SchemaValidator.validate(null, schema, null, element), element);

		schema.setType(Type.of(SchemaType.NULL));

		assertThrows(SchemaValidationException.class, () -> SchemaValidator.validate(null, schema, null, element));

	}

	@Test
	void schemaValidatorForObjectWhenTypeMissing() {

		JsonObject defaultValue = new JsonObject();
		defaultValue.addProperty("value", 123);

		Schema schema = Schema.ofObject("testSchema")
		        .setProperties(Map.of("intType", new Schema()));

		assertEquals(defaultValue, SchemaValidator.validate(null, schema, null, defaultValue));
	}

	@Test
	void schemaValidatorForStringWhenTypeMissing() {

		JsonObject defaultValue = new JsonObject();
		defaultValue.addProperty("value", "surendhar.s");

		Schema schema = Schema.ofObject("testSchema")
		        .setProperties(Map.of("stringType", new Schema()));

		assertEquals(defaultValue, SchemaValidator.validate(null, schema, null, defaultValue));
	}
}

File: kirun-js/__tests__/engine/json/schema/validator/SchemaValidatorTest.ts
import { AdditionalType, HybridRepository, Repository } from '../../../../../src';
import { Schema } from '../../../../../src/engine/json/schema/Schema';
import { SchemaType } from '../../../../../src/engine/json/schema/type/SchemaType';
import { TypeUtil } from '../../../../../src/engine/json/schema/type/TypeUtil';
import { SchemaValidationException } from '../../../../../src/engine/json/schema/validator/exception/SchemaValidationException';
import { SchemaValidator } from '../../../../../src/engine/json/schema/validator/SchemaValidator';
import { KIRunSchemaRepository } from '../../../../../src/engine/repository/KIRunSchemaRepository';

const repo = new KIRunSchemaRepository();

test('Schema Validator Test 1', async () => {
    let schema: Schema = new Schema().setType(TypeUtil.of(SchemaType.INTEGER));

    expect(await SchemaValidator.validate([], schema, repo, 2)).toBe(2);

    let obj = { name: 'shagil' };
    let objSchema: Schema = Schema.ofObject('testObj')
        .setProperties(new Map<string, Schema>([['name', Schema.ofString('name')]]))
        .setRequired(['name']);
    expect(await SchemaValidator.validate([], objSchema, repo, obj)).toBe(obj);

    expect(SchemaValidator.validate([], objSchema, repo, { name: 123 })).rejects.toThrow(
        'Value {"name":123} is not of valid type(s)\nValue 123 is not of valid type(s)\n123 is not String',
    );

    // expect(await SchemaValidator.validate([], schema, repo, 2.5)).toThrowError(new SchemaValidationException('', '2.5 is not a number of type Integer'));
});

test('Schema validation when ref of ref', async () => {
    const locationSchema = Schema.from({
        name: 'Location',
        namespace: 'Test',
        type: 'OBJECT',
        properties: {
            url: { name: 'url', type: 'STRING' },
        },
        required: ['url'],
    });

    const urlParamsSchema = Schema.ofObject('UrlParameters')
        .setNamespace('Test')
        .setAdditionalProperties(new AdditionalType().setSchemaValue(Schema.ofRef(`Test.Location`)))
        .setDefaultValue({});

    const testSchema = Schema.ofObject('TestSchema')
        .setNamespace('Test')
        .setAdditionalProperties(
            new AdditionalType().setSchemaValue(Schema.ofRef(`Test.UrlParameters`)),
        )
        .setDefaultValue({});

    const schemaMap = new Map([
        ['Location', locationSchema],
        ['UrlParameters', urlParamsSchema],
        ['TestSchema', testSchema],
    ]);

    class X implements Repository<Schema> {
        async find(namespace: string, name: string): Promise<Schema | undefined> {
            if (namespace !== 'Test') return undefined;
            return schemaMap.get(name);
        }

        async filter(name: string): Promise<string[]> {
            return Array.from(schemaMap.values())
                .map((e) => e!.getFullName())
                .filter((e) => e.toLowerCase().indexOf(name.toLowerCase()) !== -1);
        }
    }

    const repo = new HybridRepository<Schema>(new X(), new KIRunSchemaRepository());
    const obj = { obj: { url: 'http://xxxxxx.com' } };
    const queryParams = {
        obj: { obj: { url: 'http://xxxxxx.com' } },
    };

    expect(
        await SchemaValidator.validate(
            undefined,
            Schema.ofRef('Test.TestSchema'),
            repo,
            queryParams,
        ),
    ).toBe(queryParams);
});

test('Schema Validator Test 2', async () => {
    const locationSchema = Schema.from({
        name: 'Location',
        namespace: 'Test',
        type: 'OBJECT',
        properties: {
            url: { name: 'url', type: 'STRING' },
        },
        required: ['url'],
    });

    const repo = new HybridRepository<Schema>(
        {
            async find(namespace, name): Promise<Schema | undefined> {
                if (namespace === 'Test' && name === 'Location') {
                    return locationSchema;
                }
                return undefined;
            },
            async filter(name): Promise<string[]> {
                return [locationSchema!.getFullName()].filter((n) =>
                    n.toLowerCase().includes(name.toLowerCase()),
                );
            },
        },
        new KIRunSchemaRepository(),
    );

    const obj = { url: 'http://xxxx.com' };

    expect(
        await SchemaValidator.validate(undefined, Schema.ofRef('Test.Location'), repo, obj),
    ).toBe(obj);
});

test('Validate null for ofAny schema', async () => {
    expect(
        await SchemaValidator.validate(
            undefined,
            Schema.ofAny('ofanyundefined'),
            undefined,
            undefined,
        ),
    ).toBe(undefined);
});

test('Schema Validator Test 3', async () => {
    const obj = { url: 'http://xxxx.com' };

    const locationSchema = Schema.from({
        name: 'Location',
        namespace: 'Test',
        type: 'OBJECT',
        properties: {
            url: { name: 'url', type: 'String' },
        },
        required: ['url'],
        defaultValue: obj,
    });

    const repo = new HybridRepository<Schema>(
        {
            async find(namespace, name): Promise<Schema | undefined> {
                if (namespace === 'Test' && name === 'Location') {
                    return locationSchema;
                }
                return undefined;
            },

            async filter(name): Promise<string[]> {
                return [locationSchema!.getFullName()].filter((n) =>
                    n.toLowerCase().includes(name.toLowerCase()),
                );
            },
        },

        new KIRunSchemaRepository(),
    );

    const obj1 = { url: 'http://yyyy.com' };

    expect(
        await SchemaValidator.validate(
            undefined,
            Schema.ofRef('Test.Location').setDefaultValue(obj1),
            repo,
            undefined,
        ),
    ).toMatchObject(obj1);
});

test('Custom message validation - 1', async () => {
    const schema = Schema.from({
        type: 'STRING',
        name: 'test',
        namespace: 'test',
        minLength: 5,
        uiHelper: {
            validationMessages: {
                minLength:
                    'Custom minLength message is not matching {value}, you cannot use this value {value}',
            },
        },
    });

    const obj = '1234';
    expect(SchemaValidator.validate(undefined, schema, repo, obj)).rejects.toThrow(
        'Custom minLength message is not matching 1234, you cannot use this value 1234',
    );
});

test('Custom message validation - 2', async () => {
    const schema = Schema.from({
        type: 'STRING',
        name: 'test',
        namespace: 'test',
        maxLength: 15,
        uiHelper: {
            validationMessages: {
                maxLength:
                    'Custom maxLength message is not matching {value}, you cannot use this value {value}',
            },
        },
    });

    const obj = '1234567890123456';
    expect(SchemaValidator.validate(undefined, schema, repo, obj)).rejects.toThrow(
        'Custom maxLength message is not matching 1234567890123456, you cannot use this value 1234567890123456',
    );
});

test('Custom message validation - 3', async () => {
    const schema = Schema.from({
        type: 'STRING',
        name: 'test',
        namespace: 'test',
        format: 'REGEX',
        pattern: '^[A-Z]+$',
        uiHelper: {
            validationMessages: {
                pattern:
                    'Custom pattern message is not matching {value}, you cannot use this value {value}',
            },
        },
    });

    const obj = 'ABCDe';
    expect(SchemaValidator.validate(undefined, schema, repo, obj)).rejects.toThrow(
        'Custom pattern message is not matching ABCDe, you cannot use this value ABCDe',
    );
});
